File: src/main/java/com/pickme/calendar/controller/CalendarController.java
package com.pickme.calendar.controller;

import java.time.YearMonth;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pickme.calendar.dto.request.GetInterviewDto;
import com.pickme.calendar.dto.request.PostInterviewDto;
import com.pickme.calendar.dto.request.PutInterviewDto;
import com.pickme.calendar.dto.response.CalendarDto;
import com.pickme.calendar.dto.response.ResponseDto;
import com.pickme.calendar.service.CalendarService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@RestController
@RequestMapping("/calendar")
@RequiredArgsConstructor
@Slf4j
@Tag(name = "Calendar", description = "면접 캘린더 API")
@ApiResponse(responseCode = "400", description = "잘못된 요청")
@ApiResponse(responseCode = "401", description = "권한 없음")
@ApiResponse(responseCode = "404", description = "면접 일정이 없음")
public class CalendarController {

	private final CalendarService calendarService;

	// 해당 사용자 면접 일정 전체 조회
	@Operation(summary = "면접 일정 전체&조건 조회", description = "면접 일정 전체 조회 & 특정 조건에 해당하는 면접 일정 조회")
	@ApiResponse(responseCode = "200", description = "조회 요청 성공",
		content = @Content(schema = @Schema(implementation = CalendarDto.class)))
	@GetMapping("/interviews")
	public ResponseEntity<?> interviewsList(HttpServletRequest request,
		@Parameter(description = "회사 이름 (필터링 조건)", example = "앙떼띠")
		@RequestParam(required = false) String name,
		@Parameter(description = "조회할 년/월 (yyyyMM 형식, 필터링 조건)", example = "2024-11")
		@RequestParam(required = false) @DateTimeFormat(pattern = "yyyy-MM") YearMonth yearMonth) {

		String clientId = (String)request.getAttribute("clientId");

		var calendarDto = calendarService.interviewsList(clientId, name, yearMonth);
		return ResponseEntity.ok(calendarDto);
	}

	// 해당 사용자의 interviewDetailId에 해당하는 면접 일정 조회
	@Operation(summary = "면접 일정 조회", description = "interviewDetailId에 해당하는 면접 일정 조회")
	@ApiResponse(responseCode = "200", description = "조회 요청 성공",
		content = @Content(schema = @Schema(implementation = GetInterviewDto.class)))
	@GetMapping("/interview")
	public ResponseEntity<?> interview(
		@Parameter(description = "면접 일정 ID", example = "27e725b8-5816-4783-a4d0-7a19e7ae4f34")
		@RequestParam String interviewDetailId) {

		var getInterviewDto = calendarService.getInterview(interviewDetailId);
		return ResponseEntity.ok(getInterviewDto);
	}

	// 면접 일정 추가
	@Operation(summary = "면접 일정 추가", description = "새로운 면접 일정 추가")
	@ApiResponse(responseCode = "200", description = "면접 일정 추가 성공",
		content = @Content(schema = @Schema(implementation = ResponseDto.class)))
	@PostMapping("/interview")
	public ResponseEntity<?> createInterviewSchedule(HttpServletRequest request,
		@RequestBody PostInterviewDto postInterviewDto) {

		String clientId = (String)request.getAttribute("clientId");

		boolean success = calendarService.registerInterviewSchedule(postInterviewDto, clientId);
		return ResponseEntity.ok(success);
	}

	// 면접 일정 삭제
	@Operation(summary = "면접 일정 삭제", description = "interviewDetailId에 해당하는 면접 일정 삭제")
	@ApiResponse(responseCode = "200", description = "면접 일정 삭제 성공",
		content = @Content(schema = @Schema(implementation = ResponseDto.class)))
	@DeleteMapping("/interview")
	public ResponseEntity<?> deleteInterviewSchedule(
		@Parameter(description = "면접 일정 ID (필터링 조건)", example = "27e725b8-5816-4783-a4d0-7a19e7ae4f34")
		@RequestParam String interviewDetailId) {

		boolean success = calendarService.deleteInterviewSchedule(interviewDetailId);
		return ResponseEntity.ok(success);
	}

	// 특정 면접 일정 수정
	@Operation(summary = "면접 일정 수정", description = "interviewDetailId에 해당하는 면접 일정 수정")
	@ApiResponse(responseCode = "200", description = "면접 일정 수정 성공",
		content = @Content(schema = @Schema(implementation = ResponseDto.class)))
	@PutMapping("/interview")
	public ResponseEntity<?> putInterviewSchedule(
		@Parameter(description = "면접 일정 ID (필터링 조건)", example = "27e725b8-5816-4783-a4d0-7a19e7ae4f34")
		@RequestParam String interviewDetailId,
		@RequestBody PutInterviewDto putInterviewDto) {

		boolean success = calendarService.putInterviewSchedule(interviewDetailId, putInterviewDto);
		return ResponseEntity.ok(success);
	}

}

File: src/main/kotlin/com/pickme/calendar/adapter/inbound/web/controller/CalendarController.kt
package com.pickme.calendar.adapter.inbound.web.controller

import com.pickme.calendar.adapter.inbound.web.api.ApiPaths
import com.pickme.calendar.adapter.inbound.web.dto.request.PostScheduleDto
import com.pickme.calendar.adapter.inbound.web.dto.request.PutScheduleDto
import com.pickme.calendar.adapter.inbound.web.dto.request.SearchQueryDto
import com.pickme.calendar.adapter.inbound.web.dto.response.ResponseDto
import com.pickme.calendar.adapter.inbound.web.dto.response.ScheduleDto
import com.pickme.calendar.adapter.inbound.web.dto.response.ScheduleIdDto
import com.pickme.calendar.adapter.inbound.web.mapper.InterviewScheduleMapper
import com.pickme.calendar.application.usecase.*
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.Valid
import org.springdoc.core.annotations.ParameterObject
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(ApiPaths.V2)
@Tag(name = "Calendar", description = "면접 캘린더 API")
@ApiResponse(responseCode = "400", description = "잘못된 요청")
@ApiResponse(responseCode = "401", description = "권한 없음")
@ApiResponse(responseCode = "404", description = "면접 일정이 없음")
class CalendarController(
    private val findSchedules: FindSchedulesUseCase,
    private val getSchedule: GetScheduleUseCase,
    private val registerSchedule: RegisterScheduleUseCase,
    private val updateSchedule: UpdateScheduleUseCase,
    private val deleteSchedule: DeleteScheduleUseCase,
    private val scheduleMapper: InterviewScheduleMapper
) {

    @Operation(summary = "면접 일정 조회", description = "scheduleId에 해당하는 면접 일정 조회")
    @ApiResponse(responseCode = "200", description = "조회 요청 성공")
    @GetMapping("/schedule")
    fun getSchedule(
        request: HttpServletRequest,
        @Parameter(description = "면접 일정 ID", example = "694d1d9462a47e4039250532")
        @RequestParam scheduleId: String
    ): ResponseEntity<ResponseDto<ScheduleDto>> {
        val clientId = request.getAttribute("clientId") as String

        val schedule = getSchedule.execute(
            GetScheduleQuery(scheduleId, clientId)
        )

        val scheduleDto = scheduleMapper.toDto(schedule)

        return ResponseEntity.ok(
            ResponseDto(true, "면접 일정 조회 성공", scheduleDto)
        )
    }

    @Operation(summary = "면접 일정 조건 조회", description = "조건에 해당하는 면접 일정 조회")
    @ApiResponse(responseCode = "200", description = "조회 요청 성공")
    @GetMapping("/schedules")
    fun searchSchedules(
        request: HttpServletRequest,
        @ParameterObject searchQueryDto: SearchQueryDto,
    ): ResponseEntity<ResponseDto<List<ScheduleDto>>> {
        val clientId = request.getAttribute("clientId") as String
        val search = scheduleMapper.toEntity(searchQueryDto)

        val found = findSchedules.execute(
            FindSchedulesQuery(search, clientId)
        )

        val scheduleDto = scheduleMapper.toDto(found)

        return ResponseEntity.ok(
            ResponseDto(true, "면접 일정 조회 성공", scheduleDto)
        )
    }

    @Operation(summary = "면접 일정 추가", description = "새로운 면접 일정 추가")
    @ApiResponse(responseCode = "200", description = "면접 일정 추가 성공")
    @PostMapping("/schedule")
    fun createSchedule(
        request: HttpServletRequest,
        @Valid @RequestBody postScheduleDto: PostScheduleDto
    ): ResponseEntity<ResponseDto<ScheduleIdDto>> {
        val clientId = request.getAttribute("clientId") as String

        val schedule = scheduleMapper.toEntity(postScheduleDto, clientId)

        val scheduleId = registerSchedule.execute(
            RegisterScheduleCommand(clientId, schedule)
        )

        return ResponseEntity.ok(
            ResponseDto(true, "면접 일정 추가 성공", ScheduleIdDto(scheduleId))
        )
    }

    @Operation(summary = "면접 일정 수정", description = "scheduleId에 해당하는 면접 일정 수정")
    @ApiResponse(responseCode = "200", description = "면접 일정 수정 성공")
    @PutMapping("/schedule")
    fun updateSchedule(
        request: HttpServletRequest,
        @Parameter(description = "면접 일정 ID", example = "694d1d9462a47e4039250532")
        @RequestParam scheduleId: String,
        @Valid @RequestBody putScheduleDto: PutScheduleDto
    ): ResponseEntity<ResponseDto<Nothing>> {
        val clientId = request.getAttribute("clientId") as String
        val changes = scheduleMapper.toEntity(putScheduleDto)

        updateSchedule.execute(
            UpdateScheduleCommand(changes, scheduleId, clientId)
        )

        return ResponseEntity.ok(
            ResponseDto(true, "면접 일정 수정 성공")
        )
    }

    // 면접 일정 삭제
    @Operation(summary = "면접 일정 삭제", description = "scheduleId에 해당하는 면접 일정 삭제")
    @ApiResponse(responseCode = "200", description = "면접 일정 삭제 성공")
    @DeleteMapping("/schedule")
    fun deleteSchedule(
        request: HttpServletRequest,
        @Parameter(description = "면접 일정 ID", example = "694d1d9462a47e4039250532")
        @RequestParam scheduleId: String
    ): ResponseEntity<ResponseDto<Nothing>> {
        val clientId = request.getAttribute("clientId") as String

        deleteSchedule.execute(
            DeleteScheduleCommand(scheduleId, clientId)
        )

        return ResponseEntity.ok(
            ResponseDto(true, "면접 일정 삭제 성공")
        )
    }
}
